controller/persistence/base_store.go

```diff
@@ -165,39 +160,136 @@ func validateEntityIds(tx *bbolt.Tx, store NameIndexedStore, field string, ids [
 	return nil
 }
 
-func UpdateRelatedRoles(tx *bbolt.Tx, entityId string, roleSymbol boltz.EntitySetSymbol,
-	linkCollection boltz.LinkCollection, new []boltz.FieldTypeAndValue, holder errorz.ErrorHolder, semanticSymbol boltz.EntitySymbol) {
-	ids, _, err := roleSymbol.GetStore().QueryIds(tx, "true")
-	holder.SetError(err)
+type roleAttributeChangeContext struct {
+	tx                    *bbolt.Tx
+	roleSymbol            boltz.EntitySetSymbol
+	linkCollection        boltz.LinkCollection
+	relatedLinkCollection boltz.LinkCollection
+	denormLinkCollection  boltz.RefCountedLinkCollection
+	errorz.ErrorHolder
+}
 
-	entityRoles := FieldValuesToIds(new)
+func EvaluatePolicy(ctx *roleAttributeChangeContext, policy Policy, roleAttributesSymbol boltz.EntitySetSymbol) {
+	policyId := []byte(policy.GetId())
 
-	for _, id := range ids {
-		roleSet := roleSymbol.EvalStringList(tx, []byte(id))
+	roleSet := ctx.roleSymbol.EvalStringList(ctx.tx, policyId)
+	roles, ids, err := splitRolesAndIds(roleSet)
+	if err != nil {
+		ctx.SetError(err)
+		return
+	}
+
+	if err := validateEntityIds(ctx.tx, ctx.linkCollection.GetLinkedSymbol().GetStore(), ctx.roleSymbol.GetName(), ids); err != nil {
+		ctx.SetError(err)
+		return
+	}
+
+	cursor := roleAttributesSymbol.GetStore().IterateIds(ctx.tx, ast.BoolNodeTrue)
+	for ; cursor.IsValid(); cursor.Next() {
+		entityId := cursor.Current()
+		entityRoleAttributes := roleAttributesSymbol.EvalStringList(ctx.tx, entityId)
+		evaluatePolicyAgainstEntity(ctx, policy.GetSemantic(), entityId, policyId, ids, roles, entityRoleAttributes)
+	}
+}
+
+func UpdateRelatedRoles(ctx *roleAttributeChangeContext, entityId []byte, newRoleAttributes []boltz.FieldTypeAndValue, semanticSymbol boltz.EntitySymbol) {
+	cursor := ctx.roleSymbol.GetStore().IterateIds(ctx.tx, ast.BoolNodeTrue)
+
+	entityRoles := FieldValuesToIds(newRoleAttributes)
+
+	for ; cursor.IsValid(); cursor.Next() {
+		policyId := cursor.Current()
+		roleSet := ctx.roleSymbol.EvalStringList(ctx.tx, policyId)
 		roles, ids, err := splitRolesAndIds(roleSet)
 		if err != nil {
-			holder.SetError(err)
+			ctx.SetError(err)
 			return
 		}
+
 		semantic := SemanticAllOf
-		if semanticSymbol != nil {
-			if _, semanticValue := semanticSymbol.Eval(tx, []byte(id)); semanticValue != nil {
-				semantic = string(semanticValue)
-			}
+		if _, semanticValue := semanticSymbol.Eval(ctx.tx, policyId); semanticValue != nil {
+			semantic = string(semanticValue)
+		}
+		evaluatePolicyAgainstEntity(ctx, semantic, entityId, policyId, ids, roles, entityRoles)
+	}
+}
+
+func (store *baseStore) updateServicePolicyRelatedRoles(ctx *roleAttributeChangeContext, entityId []byte, newRoleAttributes []boltz.FieldTypeAndValue) {
```

Comment:
Seems weird for this to be in a file called `base_store.go`

Reply:
Moved the funcs that used by multiple policy related stores to policy_common.go